api-service/cloudify_api/routers/audit.py

```diff
@@ -77,100 +140,98 @@ async def inject_audit_logs(logs: List[InsertLog],
-                    execution_id: Optional[str] = None,
-                    since: Optional[datetime] = None) -> Select:
-    stmt = select(models.AuditLog).order_by('created_at')
-    if creator_name is not None:
-        stmt = stmt.where(models.AuditLog.creator_name == creator_name)
-    if execution_id is not None:
-        stmt = stmt.where(models.AuditLog.execution_id == execution_id)
-    if since is not None:
-        stmt = stmt.where(models.AuditLog.created_at >= since)
-    return stmt
-
-
 @router.get("/stream")
 async def stream_audit_log(request: Request,
-                           creator_name: Optional[str] = None,
-                           execution_id: Optional[str] = None,
-                           since: Optional[datetime] = None
+                           creator_name: str | None = None,
+                           execution_id: str | None = None,
+                           since: datetime | None = None
                            ) -> StreamingResponse:
     request.app.logger.debug("Handling stream_audit_log request")
     queue = asyncio.Queue()
     request.app.listener.attach_queue(NOTIFICATION_CHANNEL, queue)
     headers = {"Content-Type": "text/event-stream"}
     return StreamingResponse(
-        _audit_log_streamer(request, queue, creator_name, execution_id, since),
+        audit_log_streamer(request, queue, creator_name, execution_id, since),
         headers=headers)
 
 
-async def _audit_log_streamer(request: Request,
-                              queue: asyncio.Queue,
-                              creator_name: Optional[str] = None,
-                              execution_id: Optional[str] = None,
-                              since: Optional[datetime] = None
-                              ) -> Sequence[bytes]:
+@router.delete("", response_model=DeletedResult)
+async def truncate_audit_log(p=Depends(TruncateParams),
+                             session=Depends(make_db_session),
+                             app=Depends(get_app)):
+    app.logger.debug("truncate_audit_log, params=(%s)", p)
+    stmt = delete(models.AuditLog)\
+        .where(models.AuditLog.created_at <= p.before)
+    if p.creator_name:
+        stmt = stmt.where(models.AuditLog.creator_name == p.creator_name)
+    if p.execution_id:
+        stmt = stmt.where(models.AuditLog.execution_id == p.execution_id)
+    return await DeletedResult.executed(session, stmt)
+
+
+def select_audit_log_query(creator_name: str | None = None,
+                           execution_id: str | None = None,
+                           since: datetime | None = None) -> Select:
+    stmt = select(models.AuditLog).order_by('created_at')
+    if creator_name is not None:
+        stmt = stmt.where(models.AuditLog.creator_name == creator_name)
+    if execution_id is not None:
+        stmt = stmt.where(models.AuditLog.execution_id == execution_id)
+    if since is not None:
+        stmt = stmt.where(models.AuditLog.created_at >= since)
+    return stmt
```

Comment:
Not necessarily for right now, but a design perspective to think about...

I think instead of functions taking/returning the query/statement (like we do in the storage-manager), a better idea is to have functions that return expressions/filters.

In this case, I'm wondering about something like:

```python
def log_list_wheres(
    creator_name: str | None = None,
    execution_id: str | None = None,
    since: datetime | None = None,
):
    filters = []
    if creator_name is not None:
        filters.append(models.AuditLog.creator_name == creator_name)
    if execution_id is not None:
        filters.append(models.AuditLog.execution_id == execution_id)
    if since is not None:
        filters.append(models.AuditLog.created_at >= since)
    return filters


@router.get("", response_model=PaginatedAuditLog)
async def list_audit_log(
    filters=Depends(log_list_wheres),
    session=Depends(make_db_session),
    app=Depends(get_app),
    p=Depends(common_parameters),
) -> PaginatedAuditLog:
    query = (
        select(models.AuditLog)
        .order_by('created_at')
        .where(*filters)
    )
    return await PaginatedAuditLog.paginated(session, query, p)
```

I think this approach would be more composable between the list, the stream, and maybe even the delete.

I'm mostly just worried that with the approach of `select_audit_log_query`, especially in future endpoints, we'll end up with a storage-manager-like big "create query" function that takes fifty parameters (filter, group by, distinct, include, ...) and has `if`s on all of them, that is somewhat ugly.